server/commands/list_objects.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil {
+		return err
+	}
+
+	resultsChan := make(chan string, q.listObjectsMaxResults)
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	if q.listObjectsDeadlineInSeconds != 0 {
+		var cancel context.CancelFunc
+		timeoutCtx, cancel = context.WithTimeout(ctx, time.Second*time.Duration(q.listObjectsDeadlineInSeconds))
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, req.StoreId, req.AuthorizationModelId, req.Type, req.Relation, req.User, req.ContextualTuples, resultsChan, errChan, resolvedChan)
+	}()
+
+	for {
+		select {
+		case <-timeoutCtx.Done():
+			return serverErrors.NewInternalError("Timeout exceeded", timeoutCtx.Err())
+		case objectID, ok := <-resultsChan:
+			if !ok {
+				return nil //channel was closed
+			}
+			if err := srv.Send(&openfgapb.StreamedListObjectsResponse{
+				ObjectId: objectID,
+			}); err != nil {
+				return serverErrors.NewInternalError("", err)
+			}
+		case genericError, ok := <-errChan:
+			if ok {
+				return serverErrors.NewInternalError("", genericError)
+			}
+		}
+	}
+}
+
+func (q *ListObjectsQuery) getUniqueObjects(ctx context.Context, storeID, targetObjectType string, ctxTuples *openfgapb.ContextualTupleKeys) ([]string, error) {
+	uniqueObjects, err := q.datastore.ListObjectsByType(ctx, storage.ListObjectsFilter{
+		StoreID:    storeID,
+		ObjectType: targetObjectType,
+	})
+	if err != nil {
+		return nil, serverErrors.NewInternalError("", err)
+	}
+
+	if ctxTuples != nil {
+		// Take the relevant objects (i.e. same type) from the contextual tuples and add them to the result
+		uniqueSet := make(map[string]bool, len(uniqueObjects))
+		for _, o := range uniqueObjects {
+			_, found := uniqueSet[o]
+			if !found {
+				uniqueSet[o] = true
+			}
+		}
+
+		for _, ctxTuple := range ctxTuples.TupleKeys {
+			objectType, _ := tuple.SplitObject(ctxTuple.Object)
+			if objectType != targetObjectType {
+				continue
+			}
+			_, found := uniqueSet[ctxTuple.Object]
+			if !found {
+				uniqueSet[ctxTuple.Object] = true
+				uniqueObjects = append(uniqueObjects, ctxTuple.Object)
+			}
+		}
+	}
+	return uniqueObjects, nil
+}
+
+func (q *ListObjectsQuery) validateInput(ctx context.Context, storeID string, targetObjectType string, authModelID string, relation string) error {
+	definition, err := q.datastore.ReadTypeDefinition(ctx, storeID, authModelID, targetObjectType)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return serverErrors.TypeNotFound(targetObjectType)
+		}
+		return err
+	}
+	_, ok := definition.Relations[relation]
+	if !ok {
+		return serverErrors.UnknownRelationWhenListingObjects(relation, targetObjectType)
+	}
+	return nil
+}
+
+func (q *ListObjectsQuery) performChecks(timeoutCtx context.Context, storeID string, authModelID string, objectType string, relation string, user string, ctxTuples *openfgapb.ContextualTupleKeys, resultsChan chan<- string, errChan chan<- error, resolvedChan chan<- struct{}) {
```

Comment:
Huge method signature. Can we please refactor to be easier to read? Maybe consider using the Option pattern or an input struct instead?